### PR TITLE
[String] Add the normalizeNewline method

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

I have needed this several times so I think it would be a nice shortcut in the component.

### CODE REVIEW

File: src/Symfony/Component/String/AbstractUnicodeString.php [hunk 1]
```diff
@@ -267,6 +267,15 @@ public function normalize(int $form = self::NFC): self
         return $str;
     }
 
+    public function normalizeNewline(string $to = "\n"): AbstractString
+    {
+        if (!\in_array($to, ["\r", "\n", "\r\n", "\u{0085}", "\u{000b}", "\u{000c}", "\u{2028}", "\u{2029}"], true)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/String/AbstractUnicodeString.php:272]: no need for this check imho (same in byte string)

**Review by nicolas-grekas — COMMENTED** [review]
If the new method is just a wrapper to do `$s->replaceMatches('/\R/', $to);`, maybe it's not worth it?

Did you spot any prior art on the topic?

File: src/Symfony/Component/String/AbstractString.php [hunk 2]
```diff
@@ -426,6 +426,13 @@ public function jsonSerialize(): string
      */
     abstract public function match(string $regexp, int $flags = 0, int $offset = 0): array;
 
+    /**
+     * @param string $to the target newline
+     *
+     * @return static
+     */
+    abstract public function normalizeNewline(string $to = "\n"): self;
```
Comment (javiereguiluz) [inline src/Symfony/Component/String/AbstractString.php:434]: Just for reference purposes: I was trying to find this method in other languages to see if naming matches, but I couldn't find this method in Python or Golang APIs. Maybe it's not that common? Maybe it's hard to implement it well?

File: src/Symfony/Component/String/AbstractUnicodeString.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (stof) [reply src/Symfony/Component/String/AbstractUnicodeString.php:272]: Looks like Python has 2 separate concepts:

- universal newlines (only `\n`, `\r\n` and `\r`)
- line boundaries used by `str.splitlines`, which are a superset

Note that Python does not use the term `newline` for the extended concept.

**Comment by fancyweb** [comment]
> If the new method is just a wrapper to do $s->replaceMatches('/\R/', $to);, maybe it's not worth it?

Yes that's a wrapper to ease the normalization of newlines because from my experience it's a common task. It's good for the DX.

File: src/Symfony/Component/String/AbstractUnicodeString.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (fancyweb) [reply src/Symfony/Component/String/AbstractUnicodeString.php:272]: I took this list from http://unicode.org/standard/reports/tr13/tr13-5.html#Definitions.

File: src/Symfony/Component/String/AbstractString.php (hunk 2 above)
Comment (fancyweb) [reply src/Symfony/Component/String/AbstractString.php:434]: I did not look for it on other APIs.

**Comment by nicolas-grekas** [comment]
Thanks for the details.

I'm sorry but I'm :+1: : as hinted by @stof in https://github.com/symfony/symfony/pull/38355#discussion_r497567717, there are two concepts mixed here:
1. normalizing CRLF/LF
2. splitting texts by lines

The first use case is about providing some OS portability.
The second goes much deeper into linguistics, Unicode has pages of specs to explain how clustering texts works.

Here, this mixes both concepts and will be very confusing in practice (it won't at first, then ppl will learn about the topic and will understand they got it wrong, for sure).

**Comment by fancyweb** [comment]
The only case I care about is normalizing CRLF/LF. Could we consider having a shortcut for replaceMatches `/\r\n|\r|\n/` then?

**Comment by nicolas-grekas** [comment]
In real world, only `str_replace("\r\n", "\n")` is needed for this. ("\r" standalone is dead since tens of years)

**Comment by ro0NL** [comment]
i see some value in `toLines(string $str): string[]` generally spoken for an API. We use it to split address line into individual address lines, at least.

It plays nice with `wrap(toLines($str))` for per-line processing (trim, truncate, etc.) also.

**Comment by nicolas-grekas** [comment]
@ro0NL see [`IntlBreakIterator`](https://php.net/class.intlbreakiterator) and [`IntlRuleBasedBreakIterator`](https://php.net/class.intlrulebasedbreakiterator), this is far from trivial. Ppl that want to break texts into smaller units had better using these to me.

**Comment by ro0NL** [comment]
i agree :+1: let's not leak text-processing concerns into the string object then.

now im thinking if a `String\Text` object is the way?

**Comment by nicolas-grekas** [comment]
Closing this PR as it doesn't look worth it.
Thanks for submitting anyway!

**Pull request closed without merging**